docs/spec/type_system.rst

```diff
@@ -4,5 +4,471 @@
 Table type system
 =================
 
-TODO
+This document explains the type system of kartothek.
 
+
+Motivation
+----------
+To understand why and how the type system was designed, we illustrate the requirements and use cases kartothek should
+cover:
+
+- **simplicity:** For the average programmer, it should be possible to understand the semantics of kartothek quickly.
+- **optimized data representation:** Data providers (humans and software) should be able to pick a memory representation
+  that is sufficient to hold the data in question (e.g. pick an 8-bit unsigned integer if you only have integer up to
+  255).
+- **lifecycle management:** A dataset may be used for many months and can be extended over this time. Also, the
+  software used with the dataset might change.
+- **efficiency:** Data written once should not be altered if not really necessary.
+- **stability:** Kartothek is made for production use. It should provide stable, predicable outputs and must not crash
+  or lead other libraries to behave in an expected way.
+- **compatibility:** Kartothek is not the only software in the stack and must play nicely with others (also see `Related
+  Type Systems`_).
+
+
+Base Type System
+----------------
+Since we think that `Arrow`_ is a solid, future-proof choice for in-memory dataframes will hopefully offer a sane
```

Comment:
for in-memory dataframes `<something is missing here>` will hopefully

Reply:
We should think about whether or not we want to refer to Apache Arrow instead of plain Arrow

cc @xhochy what's your official stance to topics like these?

Reply:
The official stance is that you should use `Apache Arrow` but common usage is that you use this in headlines and once in your text and all following mentions are then simply `Arrow`.

Reply:
fixed both